gpio/__init__.py

```diff
@@ -0,0 +1,280 @@
+# -*- coding: utf-8 -*-
+__version__ = '1.0.0'
+
+from threading import Lock
+import os
+
+
+_export_lock = Lock()
+_open_pins = {}
+
+
+GPIO_ROOT = '/sys/class/gpio'
+GPIO_EXPORT = os.path.join(GPIO_ROOT, 'export')
+GPIO_UNEXPORT = os.path.join(GPIO_ROOT, 'unexport')
+FMODE = 'w+'  # w+ overwrites and truncates existing files
+IN, OUT = 'in', 'out'
+LOW, HIGH = 0, 1
+BCM = None  # Exists for RPi.GPIO compatibility
+
+
+class GPIOPin(object):
+    """Handle pin state.
+
+    Keeps track of file nodes and functions related to a pin.
+
+    Args:
+        pin (int): the GPIO pin to set up
+    """
+    def __init__(self, pin, mode=None, initial=LOW, active_low=None):
+        existing = _open_pins.get(pin)
+        if existing:
+            existing.cleanup()

```

Comment:
I don't think you want to do this on class `__init__`. Init should just be a way to "create" the data, it should not perform logic. The only thing you might have it do is throw an error saying it is "not setup" and directing hte user to use `GPIOPin.setup`

You should have a `@classmethod` called `setup` which does any setup logic you need. I think throwing an error if the pin already exists is better behavior.

All the logic around setting up the pin should be done in the `setup` classmethod.

Reply:
There should also be a `@static` called `already_opened` or something that simply returns the `GPIOPin` instance that is already open for a particular pin.

Reply:
I backtracked a bit myself. I think that if you ensure singleton instances then doing the setup in `__init__` is probably fine.